### PR TITLE
Enable additional tensor types in Gloo backend

### PR DESCRIPTION
**Pull request by myleott**
It seems to work with byte/half/char/int/long tensors:

```
import torch.distributed as dist
dist.init_process_group(init_method='tcp://127.0.0.1:23456', rank=RANK, world_size=2, backend='gloo')
# rank 0:
x = (torch.rand(10) * 10).byte().cuda()
torch.distributed.broadcast(x, src=0)
# rank 1:
y = torch.zeros(10).byte().cuda()
torch.distributed.broadcast(y, src=0)
```

Some relevant context in #2699.

### CODE REVIEW

File: torch/lib/THD/base/data_channels/DataChannelGloo.cpp
Comment (apaszke) [inline torch/lib/THD/base/data_channels/DataChannelGloo.cpp]: The type we’re using for ByteTensors is uint8_t. Can you please update that?

**Comment by myleott** [comment]
Actually, I'll add Char as well.

**Comment by apaszke** [comment]
Hmm we can't merge this if gloo doesn't have an explicit overload for uint8_t. We should send a PR to gloo first, update the submodule, and only then land this

**Comment by myleott** [comment]
Yep, working on the Gloo overload. Unfortunately Gloo needs to wrap a NCCL datatype [1], but there's no uint8 NCCL 1 datatype [2]. Still investigating...

[1] https://github.com/facebookincubator/gloo/blob/master/gloo/nccl/nccl.cu#L126-L130
[2] https://github.com/NVIDIA/nccl/blob/master/src/nccl.h#L99

**Comment by colesbury** [comment]
It seems like we should just use the `ncclChar` type if that's the only one available. I'm not sure if the fudging of types should happen in our call to Gloo with `int8_t` or Gloo's call to NCCL with `uint8_t` -> `char`.

FWIW, we map ByteTensor to ncclChar in our internal calls to NCCL:
https://github.com/pytorch/pytorch/blob/23fc2b7e06b9c539f783eb91133b1c521cb4ff9b/torch/csrc/cuda/nccl.cpp#L74

**Comment by ezyang** [comment]
CI failure is legit:

```
11:54:13 ImportError: /opt/conda/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so: undefined symbol: _ZN4gloo28CudaAllreduceHalvingDoublingIhNS_17CudaHostWorkspaceIhEEE3runEv
```

**Comment by myleott** [comment]
Yep, I've added the Gloo overload here: https://github.com/facebookincubator/gloo/commit/9b2c046e5f7d4a8ec61598d382838a8f6867a1d4.

@apaszke, how do I go about updating the submodule?

**Comment by apaszke** [comment]
@myleott Just go the the `gloo` directory, run `git pull`, go back to the PyTorch repo and commit

**Review by apaszke — APPROVED** [review]

**Pull request merged**
